micrometer-tracing/src/main/java/io/micrometer/tracing/handler/RevertingScope.java

```diff
@@ -64,4 +70,53 @@ public int hashCode() {
         return Objects.hash(tracingContext, currentScope, previousScope);
     }
 
+    static RevertingScope maybeWithBaggage(Tracer tracer, TracingContext tracingContext,
+            @Nullable TraceContext newContext, RevertingScope revertingScopeForSpan,
+            Scope previousScopeOnThisObservation) {
+        RevertingScope revertingScope = revertingScopeForSpan;
+        ContextView context = tracingContext.getContext();
+        if (context == null) {
+            return revertingScope;
+        }
+        List<KeyValue> baggageKeyValues = matchingBaggageKeyValues(tracer, context);
+        if (baggageKeyValues.isEmpty()) {
+            return revertingScope;
+        }
+        ArrayDeque<BaggageInScope> scopes = startBaggageScopes(tracer, newContext, baggageKeyValues);
+        return new RevertingScope(tracingContext, () -> {
+            for (BaggageInScope scope : scopes) {
+                scope.close();
+            }
+            revertingScope.close();
+        }, previousScopeOnThisObservation);
+    }
+
+    private static ArrayDeque<BaggageInScope> startBaggageScopes(Tracer tracer, TraceContext newContext,
+            List<KeyValue> baggageKeyValues) {
+        ArrayDeque<BaggageInScope> scopes = new ArrayDeque<>();
+        for (KeyValue keyValue : baggageKeyValues) {
+            if (newContext != null) {
+                scopes.addFirst(tracer.createBaggageInScope(newContext, keyValue.getKey(), keyValue.getValue()));
+            }
+            else {
+                scopes.addFirst(tracer.createBaggageInScope(keyValue.getKey(), keyValue.getValue()));
+            }
+        }
+        return scopes;
+    }
+
+    private static List<KeyValue> matchingBaggageKeyValues(Tracer tracer, ContextView context) {
+        List<String> lowerCaseRemoteFields = new ArrayList<>();
```

Comment:
Good to make it a `Set` for being used with `contains` for small performance optimization.

Reply:
Thanks, I've applied both suggestions